src/libaktualizr/primary/aktualizr_test.cc

```diff
@@ -807,26 +799,93 @@ TEST(Aktualizr, InstallationFailure) {
     ASSERT_TRUE(
         storage->loadInstalledVersions(primary_ecu_id, &installed_versions, &current_version, &pending_version));
 
-    ASSERT_TRUE(installed_versions.empty());
-    ASSERT_EQ(pending_version, SIZE_MAX);
-    ASSERT_EQ(current_version, SIZE_MAX);
+    EXPECT_TRUE(installed_versions.empty());
+    EXPECT_EQ(pending_version, SIZE_MAX);
+    EXPECT_EQ(current_version, SIZE_MAX);
+
+    aktualizr.UptaneCycle();
+    aktualizr.uptane_client_->completeInstall();
+
+    EXPECT_TRUE(event_hdlr.checkReceivedEvents(expected_event_order));
+    EXPECT_FALSE(aktualizr.uptane_client_->hasPendingUpdates());
+    EXPECT_TRUE(http_server_mock->checkReceivedReports(expected_report_order));
+    EXPECT_FALSE(http_server_mock->wasInstallSuccessful(primary_ecu_id));
+    EXPECT_TRUE(http_server_mock->wasInstallSuccessful(secondary_ecu_id));
+
+    data::InstallationResult dev_installation_res;
+    std::string report;
+    std::string correlation_id;
+
+    // `device_installation_result` and `ecu_installation_results` are cleared
+    // at UptaneCycle()->putManifest() once a device manifest is successfully sent to a server
+    EXPECT_FALSE(storage->loadDeviceInstallationResult(&dev_installation_res, &report, &correlation_id));
+
+    std::vector<std::pair<Uptane::EcuSerial, data::InstallationResult>> ecu_installation_res;
+    EXPECT_FALSE(storage->loadEcuInstallationResults(&ecu_installation_res));
+    EXPECT_EQ(ecu_installation_res.size(), 0);
+
+    ASSERT_TRUE(
+        storage->loadInstalledVersions(primary_ecu_id, &installed_versions, &current_version, &pending_version));
+    // it says that no any installed version found,
+    // which is, on one hand is correct since installation of the found update failed hence nothing was installed,
+    // on the other hand some version should have been installed prior to the failed update
+    EXPECT_EQ(installed_versions.size(), 0);
+    EXPECT_EQ(current_version, SIZE_MAX);
+    EXPECT_EQ(pending_version, SIZE_MAX);
+
+    fiu_disable("fake_package_install");
+  }
+
+  // primary and secondary failure
+  {
+    TemporaryDirectory temp_dir;
+    auto http_server_mock = std::make_shared<HttpInstallationFailed>(temp_dir.Path());
+    Config conf = UptaneTestCommon::makeTestConfig(temp_dir, http_server_mock->tls_server);
+    conf.bootloader.reboot_sentinel_dir = temp_dir.Path();
+    auto storage = INvStorage::newStorage(conf.storage);
+
+    fiu_init(0);
+    fault_injection_enable("fake_package_install", 1, "PRIMFAIL", 0);
```

Comment:
Maybe, makes sense to define variable for "PRIMFAIL" and "SECFAIL" as it used at three places.

Reply:
frankly, I think it won't help as building `primary_hw:PRIMFAIL|secondary_hw:SECFAIL` will be pretty verbose. I prefer to have readable than concise testing code (and it would actually increase code size probably).